drivers/accel/adxl313/iio_adxl313.c

```diff
@@ -0,0 +1,1200 @@
+static int adxl313_iio_read_range_avail(void *dev, char *buf,
+					uint32_t len, const struct iio_ch_info *channel, intptr_t priv);
+static int adxl313_iio_read_scale_avail(void *dev, char *buf,
+					uint32_t len, const struct iio_ch_info *channel, intptr_t priv);
+static int adxl313_iio_read_samples(void* dev, int* buff, uint32_t samples);
+static int adxl313_iio_update_channels(void* dev, uint32_t mask);
+static int32_t adxl313_trigger_handler(struct iio_device_data *dev_data);
+// /******************************************************************************/
+// /************************ Variable Declarations ******************************/
+// /******************************************************************************/
+
+static struct iio_attribute adxl313_iio_accel_attrs[] = {
+	{
+		.name  = "calibbias",
+		.show  = adxl313_iio_read_calibbias,
+		.store = adxl313_iio_write_calibbias,
+	},
+	{
+		.name = "raw",
+		.show = adxl313_iio_read_raw,
+	},
+	{
+		.name   = "sampling_frequency",
+		.shared = IIO_SHARED_BY_TYPE,
+		.show   = adxl313_iio_read_sampling_freq,
+		.store  = adxl313_iio_write_sampling_freq,
+	},
+	{
+		.name   = "sampling_frequency_available",
+		.shared = IIO_SHARED_BY_TYPE,
+		.show   = adxl313_iio_read_samp_freq_avail,
+	},
+	{
+		.name   = "range",
+		.shared = IIO_SHARED_BY_TYPE,
+		.show   = adxl313_iio_read_range,
+		.store  = adxl313_iio_write_range,
+	},
+	{
+		.name   = "range_available",
+		.shared = IIO_SHARED_BY_TYPE,
+		.show   = adxl313_iio_read_range_avail,
+	},
+	{
+		.name   = "scale",
+		.shared = IIO_SHARED_BY_TYPE,
+		.show   = adxl313_iio_read_scale,
+		.store  = adxl313_iio_write_scale,
+	},
+	{
+		.name   = "scale_available",
+		.shared = IIO_SHARED_BY_TYPE,
+		.show   = adxl313_iio_read_scale_avail,
+	},
+	END_ATTRIBUTES_ARRAY
+};
+
+static struct scan_type adxl313_iio_accel_scan_type = {
+	.sign = 's',
+	.realbits = 13,
+	.storagebits = 16,
+	.shift = 0,
+	.is_big_endian = false
+};
+
+#define ADXL313_ACCEL_CHANNEL(index, reg, axis) { \
+ 	.ch_type = IIO_ACCEL,                         \
+ 	.channel = index,                             \
+ 	.address = reg,                               \
+ 	.modified = true,                             \
+ 	.channel2 = IIO_MOD_##axis,                   \
+ 	.scan_type = &adxl313_iio_accel_scan_type,    \
+ 	.scan_index = index,                          \
+ 	.attributes = adxl313_iio_accel_attrs,        \
+ 	.ch_out = false                               \
+ }
+
+static struct iio_channel adxl313_channels[] = {
+	ADXL313_ACCEL_CHANNEL(0, chan_x, X),
+	ADXL313_ACCEL_CHANNEL(1, chan_y, Y),
+	ADXL313_ACCEL_CHANNEL(2, chan_z, Z),
+};
+
+static struct iio_device adxl313_iio_dev = {
+	.num_ch = NO_OS_ARRAY_SIZE(adxl313_channels),
+	.channels = adxl313_channels,
+	.pre_enable = (int32_t (*)())adxl313_iio_update_channels,
+	.trigger_handler = (int32_t (*)())adxl313_trigger_handler,
+	.read_dev = (int32_t (*)())adxl313_iio_read_samples,
```

Comment:
I'm curious if you have tested this functionality. To me it does not look like it would work seeing how you have not defined on application level a read buffer which would store the data.
You can check if this is working by using the plot functionality in IIO Osc.

Reply:
The functionality has been tested. However, the variant that includes the buffer is the right one, thank you for pointing that out.